### PR TITLE
DEV-801: Data cleaner improvements part II

### PR DESCRIPTION
**Pull request by kuuurt**

### CODE REVIEW

File: mediaire_toolbox/data_cleaner.py [hunk 1]
```diff
@@ -40,159 +52,244 @@ def __init__(self, folder, max_folder_size, max_data_seconds,
...
+        removed = []
+        scan_dirs = [entry for entry in os.scandir(base_folder) if entry.is_dir()]
+        for entry in scan_dirs:
+            removed += DataCleaner.remove_empty_folders(
+                os.path.join(base_folder, entry.name)
+            )
         return removed
 
+    def clean_up(self, dry_run=False,
+                 whitelist=None, blacklist=None):
```
[the remark this answers is not in the document]
Comment (kuuurt) [reply mediaire_toolbox/data_cleaner.py]: The idea is to add runnning spm_folder in runtime. i.e there is a transaction running and using some files from the folder. Then we could add the processing folder to the whitelist so it doesnt get deleted

File: mediaire_toolbox/data_cleaner.py [hunk 2]
```diff
@@ -40,159 +52,244 @@ def __init__(self, folder, max_folder_size, max_data_seconds,
...
+                filelist, self.max_data_seconds, whitelist, blacklist
+            )
+
+        if self.max_folder_size_bytes > 0:
+            current_size = self._sum_filestat_list(filelist)
+            reduce_size = current_size - self.max_folder_size_bytes
+            for i in range(len(self.priority_list) + 1):
+                if blacklist:
+                    priority_black_list = (
+                        blacklist + self._merge_lists(self.priority_list[::-1][:i]))
```
[the remark this answers is not in the document]
Comment (kuuurt) [reply mediaire_toolbox/data_cleaner.py]: This is a priority black list. With the priority ascending, it means that the last on in the priority list should be deleted first. But that isnt documented very well, will add that

File: mediaire_toolbox/data_cleaner.py [hunk 3]
```diff
@@ -40,159 +52,244 @@ def __init__(self, folder, max_folder_size, max_data_seconds,
...
+    def _remove_from_file_list(filelist, removed_index_list):
+        """Inplace Remove the indexes of a list given the list of indexes."""
+        shift_counter = 0
+        for i in removed_index_list:
+            del filelist[i-shift_counter]
+            shift_counter += 1
+
+    @staticmethod
+    def _merge_lists(input_list):
+        """Squash a list of lists to one list"""
```
[the remark this answers is not in the document]
Comment (kuuurt) [reply mediaire_toolbox/data_cleaner.py]: squashes a two dimensional list to one dimension. this turns [["*.nii", ".db"], ["*.dcm"]] to ["*.nii", "*.db", "*.dcm"]

[the remark this answers is not in the document]
Comment (kuuurt) [reply mediaire_toolbox/data_cleaner.py]: Ah yes, you right! Sloppy of me, I will fix that

File: mediaire_toolbox/data_cleaner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pereferrera) [reply mediaire_toolbox/data_cleaner.py]: ah ok, very good idea

**Comment by pereferrera** [comment]
Tried to reproduce a scenario like this:
``` 
cleaner = DataCleaner('/home/pere/workspace/container_orchestration/shared_data/mdbrain/dicoms', 1024, -1, whitelist=['orig_sT2.nii', 'segmLesions.nii'], priority_list=['*.dcm', '*.nii'])

out = cleaner.clean_up(dry_run=True)
```
However I get that segmLesions and sT2 files would also be deleted?

I also have a question: do we still delete first old subfolders? Or is this business logic of iterating over subfolders from oldest to newest not applied anymore? Not sure if we would really need it after the priority list, though. We'd have to discuss it.

**Comment by kuuurt** [comment]
No, we do not delete first old subfolders, instead we delete first old folders. So I just threw the whole business logic of folders because i though it was a bit obsolete with the priority list

**Pull request merged**